src/Common/PipeWriterStream.cs

```diff
@@ -57,17 +57,28 @@ public override void Write(byte[] buffer, int offset, int count)
 
         public override Task WriteAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
         {
-            Write(buffer, offset, count);
-            return Task.CompletedTask;
+            return WriteCoreAsync(buffer.AsMemory(offset, count), cancellationToken).AsTask();
         }
 
 #if NETCOREAPP2_1
         public override ValueTask WriteAsync(ReadOnlyMemory<byte> source, CancellationToken cancellationToken = default)
         {
-            _pipeWriter.Write(source.Span);
+            return WriteCoreAsync(source, cancellationToken);
+        }
+#endif
+
+        private ValueTask WriteCoreAsync(ReadOnlyMemory<byte> source, CancellationToken cancellationToken = default)
+        {
             _length += source.Length;
+            var task = _pipeWriter.WriteAsync(source);
+            if (!task.IsCompletedSuccessfully)
+            {
+                return WriteSlowAsync(task);
+            }
+
             return default;
+
+            async ValueTask WriteSlowAsync(ValueTask<FlushResult> flushTask) => await flushTask;
```

Comment:
Minor style nit: If you put this at the top, it ensures you can't accidentally capture locals from the enclosing function because nothing has been defined yet. Can't prevent arguments, but it helps.

Source: https://sharplab.io/#v2:CYLg1APgAgDABFAjAbgLACgoGYECY4DCcA3hnOQjkvAQIYAOALgK4BOApgBQCUJZFAgG61WcAGYB7CXAC8cAEQAjEfLToBAqAHY4AEQkB1CawDWPNRor9LAZUasAlgDsA5nsPGzvGQD5xUgDoAFQkAVXp6dlZzazgAX1jY7AREeAA5CTomNi5eUnVLODtHV3cjUx5ZP0kJYLCIqJj0WKERf2k5JRULQu0yzybCuFiE9DigA=